File: 02322. Minimum Score After Removals on a Tree/solution.go
/**
 * Problem: 2322. Minimum Score After Removals on a Tree
 *
 * Difficulty: Hard
 *
 * Language: Golang
 *
 * Performance: Runtime - 6 ms (Beats 100%)
 */

func calc(part1, part2, part3 int) int {
	// Calculate the minimum score after removing two edges
	return max(part1, max(part2, part3)) - min(part1, min(part2, part3))
}

func minimumScore(nums []int, edges [][]int) int {
	// Number of nodes in the tree
	n := len(nums)
	// Adjacency list to represent the tree
	adj := make([][]int, n)

	// Build the adjacency list from the edges
	for _, e := range edges {
		adj[e[0]] = append(adj[e[0]], e[1])
		adj[e[1]] = append(adj[e[1]], e[0])
	}

	// sum[i] stores the XOR sum of the subtree rooted at node i
	sum := make([]int, n)
	// in[i] stores the entry time of node i in DFS traversal
	in := make([]int, n)
	// out[i] stores the exit time of node i in DFS traversal
	out := make([]int, n)
	// cnt is a counter for DFS traversal time
	cnt := 0

	// dfs traverses the tree to calculate subtree XOR sums and entry/exit times
	var dfs func(node, parent int)
	dfs = func(node, parent int) {
		in[node] = cnt // mark entry time
		cnt++
		sum[node] = nums[node] // initialize subtree XOR sum with node value

		for _, neighbor := range adj[node] {
			if neighbor == parent {
				continue // skip parent to avoid cycling
			}

			dfs(neighbor, node) // DFS on child
			sum[node] ^= sum[neighbor] // accumulate XOR sum from child subtree
		}

		out[node] = cnt // mark exit time
	}

	// Start DFS from root node 0 with no parent
	dfs(0, -1)

	// Initialize result with maximum possible integer value
	res := math.MaxInt32
	
	// Try all pairs of nodes (u, v) to remove two edges and partition the tree
	for u := 1; u < n; u++ {
		for v := u + 1; v < n; v++ {
			// Check if v is in the subtree of u
			if in[v] > in[u] && in[v] < out[u] {
				// Partition: root excluding u, u excluding v, v subtree
				res = min(res, calc(sum[0]^sum[u], sum[u]^sum[v], sum[v]))
			} else if in[u] > in[v] && in[u] < out[v] {
				// Partition: root excluding v, v excluding u, u subtree
				res = min(res, calc(sum[0]^sum[v], sum[v]^sum[u], sum[u]))
			} else {
				// Partition: root excluding u and v, u subtree, v subtree
				res = min(res, calc(sum[0]^sum[u]^sum[v], sum[u], sum[v]))
			}
		}
	}
	
	// Return the minimum score found
	return res
}

File: 02322. Minimum Score After Removals on a Tree/solution.js
/**
 * Problem: 2322. Minimum Score After Removals on a Tree
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 96 ms (Beats 100%)
 */

/**
 * Returns the difference between the largest and smallest of three numbers
 *
 * @param {number} part1 - The first part of the calculation
 * @param {number} part2 - The second part of the calculation
 * @param {number} part3 - The third part of the calculation
 *
 * @returns {number} - The maximum difference between the largest and smallest of the three parts
 */
const calc = (part1, part2, part3) =>
  Math.max(part1, Math.max(part2, part3)) -
  Math.min(part1, Math.min(part2, part3))

/**
 * Finds the minimum score after removing two edges from a tree
 *
 * @param {number[]} nums - Node values
 * @param {number[][]} edges - Tree edges [u, v]
 *
 * @returns {number} - Minimum score
 */
const minimumScore = (nums, edges) => {
  // Number of nodes in the tree
  const nodeCount = nums.length

  // Adjacency list to represent the tree
  const adjacencyList = Array.from({ length: nodeCount }, () => [])

  // Build the adjacency list from the edges
  for (const [u, v] of edges) {
    adjacencyList[u].push(v)
    adjacencyList[v].push(u)
  }

  // Arrays to store subtree XOR sums, entry times, and exit times for each node
  const subtreeXorSum = new Array(nodeCount).fill(0),
    entryTime = new Array(nodeCount).fill(0),
    exitTime = new Array(nodeCount).fill(0)
  let timeCounter = 0

  // Helper function to perform DFS and calculate subtree XOR sums
  const dfs = (currentNode, parentNode) => {
    // Record entry time for current node
    entryTime[currentNode] = timeCounter++
    // Initialize subtree XOR sum with current node's value
    subtreeXorSum[currentNode] = nums[currentNode]

    // Traverse all adjacent nodes
    for (const neighbor of adjacencyList[currentNode]) {
      // Skip the parent node to avoid revisiting
      if (neighbor === parentNode) continue

      // Recursively visit child nodes
      dfs(neighbor, currentNode)
      // Update subtree XOR sum with child's subtree XOR sum
      subtreeXorSum[currentNode] ^= subtreeXorSum[neighbor]
    }

    // Record exit time for current node
    exitTime[currentNode] = timeCounter
  }

  // Start DFS traversal from root node (node 0)
  dfs(0, -1)

  // Initialize result with infinity for minimum score comparison
  let minScore = Infinity

  // Iterate over all pairs of nodes to consider removing two edges
  for (let firstNode = 1; firstNode < nodeCount; firstNode++) {
    for (let secondNode = firstNode + 1; secondNode < nodeCount; secondNode++) {
      //
      // Case 1: secondNode is in the subtree of firstNode
      if (
        entryTime[secondNode] > entryTime[firstNode] &&
        entryTime[secondNode] < exitTime[firstNode]
      )
        minScore = Math.min(
          minScore,
          calc(
            subtreeXorSum[0] ^ subtreeXorSum[firstNode],
            subtreeXorSum[firstNode] ^ subtreeXorSum[secondNode],
            subtreeXorSum[secondNode]
          )
        )
      //
      // Case 2: firstNode is in the subtree of secondNode
      else if (
        entryTime[firstNode] > entryTime[secondNode] &&
        entryTime[firstNode] < exitTime[secondNode]
      )
        minScore = Math.min(
          minScore,
          calc(
            subtreeXorSum[0] ^ subtreeXorSum[secondNode],
            subtreeXorSum[secondNode] ^ subtreeXorSum[firstNode],
            subtreeXorSum[firstNode]
          )
        )
      //
      // Case 3: firstNode and secondNode are in different subtrees
      else
        minScore = Math.min(
          minScore,
          calc(
            subtreeXorSum[0] ^
              subtreeXorSum[firstNode] ^
              subtreeXorSum[secondNode],
            subtreeXorSum[firstNode],
            subtreeXorSum[secondNode]
          )
        )
    }
  }

  // Return the minimum score found
  return minScore
}
